sanic/asgi.py

```diff
@@ -132,12 +132,18 @@ async def create(
         instance.response = None
         setattr(instance.transport, "add_task", sanic_app.loop.create_task)
 
-        headers = Header(
-            [
-                (key.decode("latin-1"), value.decode("latin-1"))
-                for key, value in scope.get("headers", [])
-            ]
-        )
+        try:
+            headers = Header(
+                [
+                    (
+                        key.decode("ASCII"),
+                        value.decode(errors="surrogateescape"),
+                    )
+                    for key, value in scope.get("headers", [])
+                ]
+            )
+        except UnicodeDecodeError:
+            BadRequest("Header names can only contain US-ASCII characters")
```

Comment:
Missing `raise`

Reply:
Also needs a test that hits this error.

Reply:
Added `raise`. Will add the test later today.

Reply:
In my experiment, raising a `BadRequest` in `asgi.py` will cause the ASGI server return 500 error, instead of 400.

Reply:
@ChihweiLHBird You are correct. I'll look into this probably next week.